codalab/apps/authenz/models.py

```diff
@@ -32,4 +41,57 @@ class ClUser(auth_models.AbstractUser):
     rabbitmq_username = models.CharField(max_length=36, null=True, blank=True)
     rabbitmq_password = models.CharField(max_length=36, null=True, blank=True)
 
+    def save(self, *args, **kwargs):
+        if self.newsletter_opt_in and self.email and self.is_active:
+            data = {
+                "email_address": self.email,
+                "status": "subscribed",
+            }
+
+            user_hash = hashlib.md5(str.lower(self.email.encode()))
+
+            # Try to update the user before creating a new user
+            update_user = requests.patch(
+                settings.MAILCHIMP_MEMBERS_ENDPOINT_NEWSLETTER + '/' + user_hash.hexdigest(),
+                auth=("", settings.MAILCHIMP_API_KEY),
+                data=json.dumps(data)
+            )
+
+            # If no user is found in mailchimp, create a new mailing list user
+            if not update_user.ok:
+                requests.post(
+                    settings.MAILCHIMP_MEMBERS_ENDPOINT_NEWSLETTER,
+                    auth=("", settings.MAILCHIMP_API_KEY),
+                    data=json.dumps(data)
+                )
+
+            # If there is no NewsletterUser for this user, create one
+            if not NewsletterUser.objects.filter(email=self.email).exists():
+                NewsletterUser.objects.create(email=self.email)
+
+        elif not self.email and self.newsletter_opt_in:
```

Comment:
Can check this first, as a standalone if statement. then, after this `if`, you can calculate your `user_hash` outside of the next `if...else` and then you only have to calc it once, so you can make a small DRY reduction there.

Reply:
Resolved in latest push